### PR TITLE
Search v3 Saga

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

Adds sagas that use entities. No searchv3 reducer!

use the `actions/searchv3/creators::search` function to create an action that will create a search result and save the results to the given keypath in entities.

### CODE REVIEW

File: shared/reducers/entities.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ export default function(state: State = initialState, action: Actions): State {
     case 'entity:replace': {
       const {keyPath, entities} = action.payload
       // $FlowIssue doesn't understand this API
-      return state.mergeIn(keyPath, entities)
+      return state.setIn(keyPath, entities)
```
Comment (MarcoPolo) [inline shared/reducers/entities.js]: Pretty sure this is what we want

Comment (chrisnojima) [inline shared/reducers/entities.js]: Mmm I don't think so. We want to replace the entities in there. Not replace the entire key path. Maybe we need 2 actions?

File: shared/constants/searchv3.js [hunk 2]
```diff
@@ -31,11 +37,183 @@ export type RowProps = {|
...
   showTrackerButton: boolean,
-
-  onShowTracker: () => void,
 |}
 
+// Keypaths - maybe these should be somewhere else?
+export type KeyPath = ['searchv3Chat'] | ['searchv3Profile']
+
+// Actions
+
```
Comment (chromakode) [inline shared/constants/searchv3.js]: :scissors: extra line?

File: shared/constants/searchv3.js [hunk 3]
```diff
@@ -31,11 +37,183 @@ export type RowProps = {|
...
+}
+
+function externalServiceToId(serviceName: string, serviceUsername: string): string {
+  return `${serviceName}-${serviceUsername}`
+}
+
+function _parseKeybaseRawResultToRow(
+  result: RawResult,
+  isFollowingOnKeybase: boolean,
+  showTrackerButton: boolean
```
Comment (chromakode) [inline shared/constants/searchv3.js]: Why make this an argument? Can this value always be determined from the `result` passed in?

File: shared/reducers/entities.js (hunk 1 above)
Comment (chromakode) [inline shared/reducers/entities.js]: Perhaps the `mergeIn` variant should be called "add" as opposed to "replace?

File: shared/searchv3/results-list/container.js [hunk 4]
```diff
@@ -5,8 +5,13 @@ import SearchResultsList from '.'
 import type {TypedState} from '../../constants/reducer'
 
 // TODO use entities
-const mapStateToProps = (state: TypedState, id: string) => {
-  return {}
+const mapStateToProps = ({entities}: TypedState, {keyPath}) => {
+  // $FlowIssue
+  const items = entities.getIn(keyPath).keySeq().toArray()
```
Comment (chromakode) [inline shared/searchv3/results-list/container.js]: Should the results be sorted by score?

**Comment by MarcoPolo** [comment]
Changed this so now you fire a search action that looks like:
`{type:'searchv3:search', payload: {term: 'marco', service: '', actionTypeToFire: 'search:fooo'}}` (Using the helper of course).

The search saga takes on that and fires an action of the type

`{type: actionTypeToFire, payload: {searchTerm, searchResults, service}}`


You can have another saga listen to your action type that you set, or just have it emit the 'chat:updateSearchResult' to have the reducer use it as is.

**Comment by keybase-ci-visdiff** [comment]
The commits 44153060a218...f1e001c17847 introduce visual changes on linux. <details><summary>:mag_right: 3 changed</summary>

 * Changed: **Search_V_3_results_List-keybase_Results_One** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-keybase_Results_One-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-keybase_Results_One-f1e001c17847.png)  
   Search_V_3_results_List-keybase_Results_One comparison
 * Changed: **Search_V_3_results_List-keybase_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-keybase_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-keybase_Results-f1e001c17847.png)  
   Search_V_3_results_List-keybase_Results comparison
 * Changed: **Search_V_3_results_List-facebook_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-facebook_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496172009255-44153060a218-f1e001c17847-linux/Search_V_3_results_List-facebook_Results-f1e001c17847.png)  
   Search_V_3_results_List-facebook_Results comparison
</summary>

File: shared/actions/searchv3/index.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import {apiserverGetWithSessionRpc} from '../../constants/types/flow-types'
+import {trim, keyBy} from 'lodash'
+import {call, put, select} from 'redux-saga/effects'
+import * as Selectors from '../../constants/selectors'
+import * as Saga from '../../util/saga'
+
+import type {SagaGenerator} from '../../constants/types/saga'
+
+function _apiSearch(searchTerm: string, service: string = '', limit: number = 20) {
+  return new Promise((resolve, reject) => {
```
Comment (chrisnojima) [inline shared/actions/searchv3/index.js]: we already have promise helpers for this generated right? https://github.com/keybase/client/blob/master/protocol/js/flow-types.js#L1082

File: shared/actions/searchv3/index.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+function* search<T>({payload: {term, service, actionTypeToFire}}: Constants.Search<T>) {
+  const searchQuery = Constants.toSearchQuery(service, term)
+  const cachedResults = yield select(Selectors.cachedSearchResults, searchQuery)
+  if (cachedResults) {
+    yield put(Creators.finishedSearch(actionTypeToFire, cachedResults, term, service))
+    return
+  }
+
+  try {
+    const searchResults = yield call(_apiSearch, trim(term), service)
```
Comment (chrisnojima) [inline shared/actions/searchv3/index.js]: _apiSearch should likely trim in case we call it from elsewhere

File: shared/actions/searchv3/index.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    const isFollowingFn = yield select(Selectors.isFollowingFnSelector)
+    const rows = searchResults.list.map((result: Constants.RawResult) => {
+      const isFollowingOnKeybase = !!result.keybase && isFollowingFn(result.keybase.username)
+      return Constants.parseRawResultToRow(result, service || 'Keybase', isFollowingOnKeybase)
```
Comment (chrisnojima) [inline shared/actions/searchv3/index.js]: we should likely not mix the following info here. The connected row should connect it so it gets updated correctly

File: shared/constants/searchv3.js [hunk 8]
```diff
@@ -31,11 +40,215 @@ export type RowProps = {|
...
+  {searchResults: List<SearchResultId>, searchTerm: string, service: SearchPlatform}
+>
+
+export type OnShowTracker = NoErrorTypedAction<'searchv3:onShowTracker', {resultId: string, keyPath: KeyPath}>
+
+// Generic so others can make their own version
+export type UpdateSearchResultsGeneric<T> = NoErrorTypedAction<T, {searchResults: List<SearchResultId>}>
+
+// Helper
+function serviceNameToSearchPlatform(serviceName: string): SearchPlatform {
```
Comment (chrisnojima) [inline shared/constants/searchv3.js]: This is the old code right, you should take this from my service filter code

File: shared/constants/searchv3.js [hunk 9]
```diff
@@ -31,11 +40,215 @@ export type RowProps = {|
...
+function toSearchQuery(serviceName: string, searchTerm: string): SearchQuery {
+  return `${serviceName}-${searchTerm}`
+}
+
+function _parseKeybaseRawResultToRow(
+  result: RawResult,
+  isFollowingOnKeybase: boolean,
+  showTrackerButton: boolean
+): RowProps {
+  if (result.keybase && result.service) {
```
Comment (chrisnojima) [inline shared/constants/searchv3.js]: we should likely move this into the saga. nothing else does this

File: shared/searchv3/results-list/container.js (hunk 4 above)
Comment (MarcoPolo) [inline shared/searchv3/results-list/container.js]: we don't do that with our old search code, and it looks like we do get a sorted list from the api

**Comment by keybase-ci-visdiff** [comment]
The commits 44153060a218...8894e0bb55fc introduce visual changes on linux. <details><summary>:mag_right: 6 changed</summary>

 * Changed: **unlock_folders-Success** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/unlock_folders-Success-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/unlock_folders-Success-8894e0bb55fc.png)  
   unlock_folders-Success comparison
 * Changed: **Tracker-Tracker_Error** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Tracker-Tracker_Error-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Tracker-Tracker_Error-8894e0bb55fc.png)  
   Tracker-Tracker_Error comparison
 * Changed: **Tracker-Platforms** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Tracker-Platforms-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Tracker-Platforms-8894e0bb55fc.png)  
   Tracker-Platforms comparison
 * Changed: **Search_V_3_results_List-keybase_Results_One** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-keybase_Results_One-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-keybase_Results_One-8894e0bb55fc.png)  
   Search_V_3_results_List-keybase_Results_One comparison
 * Changed: **Search_V_3_results_List-keybase_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-keybase_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-keybase_Results-8894e0bb55fc.png)  
   Search_V_3_results_List-keybase_Results comparison
 * Changed: **Search_V_3_results_List-facebook_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-facebook_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496181759386-44153060a218-8894e0bb55fc-linux/Search_V_3_results_List-facebook_Results-8894e0bb55fc.png)  
   Search_V_3_results_List-facebook_Results comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 44153060a218...b33c91999990 introduce visual changes on linux. <details><summary>:mag_right: 3 changed</summary>

 * Changed: **Search_V_3_results_List-keybase_Results_One** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-keybase_Results_One-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-keybase_Results_One-b33c91999990.png)  
   Search_V_3_results_List-keybase_Results_One comparison
 * Changed: **Search_V_3_results_List-keybase_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-keybase_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-keybase_Results-b33c91999990.png)  
   Search_V_3_results_List-keybase_Results comparison
 * Changed: **Search_V_3_results_List-facebook_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-facebook_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183018850-44153060a218-b33c91999990-linux/Search_V_3_results_List-facebook_Results-b33c91999990.png)  
   Search_V_3_results_List-facebook_Results comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 44153060a218...e8243e13cec7 introduce visual changes on linux. <details><summary>:mag_right: 3 changed</summary>

 * Changed: **Search_V_3_results_List-keybase_Results_One** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-keybase_Results_One-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-keybase_Results_One-e8243e13cec7.png)  
   Search_V_3_results_List-keybase_Results_One comparison
 * Changed: **Search_V_3_results_List-keybase_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-keybase_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-keybase_Results-e8243e13cec7.png)  
   Search_V_3_results_List-keybase_Results comparison
 * Changed: **Search_V_3_results_List-facebook_Results** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-facebook_Results-44153060a218.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496183413685-44153060a218-e8243e13cec7-linux/Search_V_3_results_List-facebook_Results-e8243e13cec7.png)  
   Search_V_3_results_List-facebook_Results comparison
</summary>

**Pull request merged**